src/Storages/IStorageDataLake.h

```diff
@@ -0,0 +1,229 @@
+    // 2. Finds out parts with latest version
+    // 3. Creates url for underlying StorageS3 enigne to handle reads
+    IStorageDataLake(
+        const StorageS3::Configuration & configuration_,
+        const StorageID & table_id_,
+        ColumnsDescription columns_,
+        const ConstraintsDescription & constraints_,
+        const String & comment,
+        ContextPtr context_,
+        std::optional<FormatSettings> format_settings_)
+        : IStorage(table_id_)
+        , base_configuration{configuration_}
+        , log(&Poco::Logger::get("Storage" + String(name) + "(" + table_id_.table_name + ")"))
+        , table_path(base_configuration.url.key)
+    {
+        StorageInMemoryMetadata storage_metadata;
+        StorageS3::updateS3Configuration(context_, base_configuration);
+
+        auto new_configuration = getAdjustedS3Configuration(context_, base_configuration, log);
+
+        if (columns_.empty())
+        {
+            columns_ = StorageS3::getTableStructureFromData(
+                new_configuration, /*distributed processing*/ false, format_settings_, context_, nullptr);
+            storage_metadata.setColumns(columns_);
+        }
+        else
+            storage_metadata.setColumns(columns_);
+
+        storage_metadata.setConstraints(constraints_);
+        storage_metadata.setComment(comment);
+        setInMemoryMetadata(storage_metadata);
+
+        s3engine = std::make_shared<StorageS3>(
+            new_configuration,
+            table_id_,
+            columns_,
+            constraints_,
+            comment,
+            context_,
+            format_settings_,
+            /* distributed_processing_ */ false,
+            nullptr);
+    }
+
+    static constexpr auto name = Name::name;
+    String getName() const override { return name; }
+
+    // Reads latest version of Lake Table
+    Pipe read(
+        const Names & column_names,
+        const StorageSnapshotPtr & storage_snapshot,
+        SelectQueryInfo & query_info,
+        ContextPtr context,
+        QueryProcessingStage::Enum processed_stage,
+        size_t max_block_size,
+        size_t num_streams) override
+    {
+        StorageS3::updateS3Configuration(context, base_configuration);
+
+        return s3engine->read(column_names, storage_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
+    }
+
+    static ColumnsDescription getTableStructureFromData(
+        StorageS3::Configuration & configuration, const std::optional<FormatSettings> & format_settings, ContextPtr ctx)
+    {
+        StorageS3::updateS3Configuration(ctx, configuration);
+        auto new_configuration = getAdjustedS3Configuration(ctx, configuration, &Poco::Logger::get("Storage" + String(name)));
+        return StorageS3::getTableStructureFromData(
+            new_configuration, /*distributed processing*/ false, format_settings, ctx, /*object_infos*/ nullptr);
+    }
+
+    static StorageS3::Configuration
```

Comment:
not sure what was the point in all those templates if we anyway use s3 explicitly?

Reply:
I get rid of S3 related things in above metadata parser, such we can easily to implement parser if we want to read the data lake tables store on other filesystem like HDFS. Buf for `IStorageDataLake` class, it's not easy to get rid of S3 related things.